airflow/models/dag.py

```diff
@@ -950,6 +990,60 @@ def clear(
         if only_running:
             tis = tis.filter(TI.state == State.RUNNING)
 
+        if include_subdags:
+            # Recursively find external tasks indicated by ExternalTaskMarker
+            instances = tis.all()
+            for ti in instances:
+                if ti.operator == ExternalTaskMarker.__name__:
```

Comment:
Isn't `ti.operator` a class? So this is basically doing `if Class() == "Class"` isn't it?

Reply:
I looked into taskinstance.py. TaskInstance.operator is a String stored in the database. It's assigned the name of the class during its construction.
```
operator = Column(String(1000))
```

However, you have a point that this line is a little awkward. Is there a better way to say "if this taskinstance is created from a instance of `ExternalTaskMarker`" ?

Reply:
`isinstance(ti.task, ExternalTaskMarker)` -- make sure that the import for this is "delayed" until it's needed, i.e. somewhere in "this" block. (not the loop, but the "if clearing these dags")

@kaxil Which of these is better from a Dag Serialization PoV? Doesn't matter right?

Reply:
Hi @ashb I tried the following:

```
isinstance(ti.task, ExternalTaskMarker)
```

Unfortunately I realized that ti.task is not set if `TaskInstance` is not created from the constructor `__init__`. 
```
E.g. the following raises `AttributeError` in this scope.
(Pdb++) p ti.task
*** AttributeError: 'TaskInstance' object has no attribute 'task'
(Pdb++) p isinstance(ti.task, ExternalTaskMarker)
*** AttributeError: 'TaskInstance' object has no attribute 'task'

This works:
(Pdb++) p ti.operator
'ExternalTaskMarker'
```

If `isinstance` is preferred, one other way do it is something like this:

```
task = self.get_task(ti.task_id)

if isinstance(task, ExternalTaskMarker):
    if recursion_depth == 0:
        # Maximum recursion depth allowed is the recursion_depth of the first
        # ExternalTaskMarker in the tasks to be cleared.
        max_recursion_depth = task.recursion_depth
    ...
```

This always gets the task object from the dag, even when it's not needed. So it's likely slightly slower than checking the value of `ti.operator`. We should only do it if there's a strong reason.

@ashb  @kaxil what do you think?

Reply:
Probably stick with it as you have it then?

Reply:
@ashb  Sure, will do. I have also moved the `import ExternalTaskMarker` inline into the `if` block like you recommended. Thank you!